File: controllers/supervisor/player.py
import json
import socket
import sys


class Frame():
    def __init__(self):
        self.time = None
        self.score = None
        self.reset_reason = None
        self.game_state = None
        self.subimages = None
        self.coordinates = None
        self.half_passed = None


class Game():
    # reset_reason
    NONE = 0
    GAME_START = 1
    SCORE_MYTEAM = 2
    SCORE_OPPONENT = 3
    GAME_END = 4
    DEADLOCK = 5
    GOALKICK = 6
    CORNERKICK = 7
    PENALTYKICK = 8
    HALFTIME = 9
    EPISODE_END = 10

    # game_state
    STATE_DEFAULT = 0
    STATE_KICKOFF = 1
    STATE_GOALKICK = 2
    STATE_CORNERKICK = 3
    STATE_PENALTYKICK = 4


class Player():
    # coordinates
    MY_TEAM = 0
    OP_TEAM = 1
    BALL = 2
    X = 0
    Y = 1
    TH = 2
    ACTIVE = 3
    TOUCH = 4

    def __init__(self):
        self.host = sys.argv[1]
        self.port = int(sys.argv[2])
        self.key = sys.argv[3]
        self.data = sys.argv[4]
        self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.socket.connect((self.host, self.port))

    def send(self, message, arguments=[]):
        message = 'aiwc.' + message + '("%s"' % self.key
        for argument in arguments:
            if isinstance(argument, str):  # string
                message += ', "%s"' % argument
            else:  # number
                message += ', %s' % argument
        message += ')'
        self.socket.sendall(message.encode())

    def receive(self):
        data = self.socket.recv(4096)
        return data.decode()

    def set_speeds(self, speeds):
        self.send('set_speeds', speeds)

    def check_frame(self, frame):  # you should override this method
        if 'reset_reason' in frame and frame['reset_reason'] == Game.GAME_END:
            return False
        return True

    def init(self, info):  # you should override this method
        print("init() method called")

    def update(self, frame): # you should override this method
        print("update() method called")

    def finish(self): # you should override this method
        print("finish() method called")

    def run(self):
        self.send('get_info')
        info = self.receive()

        self.init(json.loads(info))

        self.send('ready')

        while True:
            frame = self.receive()
            if frame and self.check_frame(json.loads(frame)): # return False if we need to quit
                self.update(json.loads(frame))
            else:
                self.finish()
                break

File: examples_new/player_random-walk/player.cpp
#include "Player.hpp"

#include <arpa/inet.h>
#include <stdio.h>
#include <sys/socket.h>
#include <unistd.h>

Player::Player(std::string host, int port, std::string key, std::string data) {
  mKey = key;
  mData = data;

  struct sockaddr_in server_addr = {0};

  // assign IP, PORT
  server_addr.sin_family = AF_INET;
  server_addr.sin_addr.s_addr = inet_addr(host.c_str();
  server_addr.sin_port = htons(port);

  // create the socket
  mConnFd = socket(AF_INET, SOCK_STREAM, 0);
  if (mConnFd == -1) {
    printf("socket creation failed...\n");
    exit(0);
  }

  // connect the client socket to server socket
  if (connect(mConnFd, (struct sockaddr*)&server_addr, sizeof(server_addr)) != 0) {
    printf("connection with the server failed...\n");
    exit(0);
  }
}

Player::~Player() {
  if (shutdown(mConnFd, SHUT_RDWR) == -1 || close(mConnFd) == -1)
    printf("Failed to shutdown connection with the server...\n");
}

void Player::sendToServer(std::string message, std::string arguments) {
  std::string toSend = "aiwc.(\"" + mKey + "\"";
  if (arguments.size() > 0)
    toSend += ',' + arguments toSend += ")";
  send(mConnFd, (void *)toSend.c_str(), sizeof(toSend.c_str()), 0);
}

void Player::receive() {
  char *buffer = new char[4096];
  memset(buffer, '0', sizeof(buffer));
  int ret = read(mConnFd, (void *)buffer, sizeof(buffer) - 1);
  if (ret > 0)
    return buffer;
  delete buffer;
  return NULL;
}

void Player::setSpeeds(std::vector speeds) {
  std::string arguments = "";
  for (unsigned i = 0; i < speeds.size(); i++)
    arguments += speeds[i] + ",";
  sendToServer("set_speeds", command);
}

bool Player::check_frame(/*frame*/) { // you should override this method
  // if "reset_reason" in frame and frame['reset_reason'] == Game.GAME_END:
  //   return false
  return true
}

void Player::init(std::string info) { // you should override this method
  printf("init() method called...\n")
}

void Player::update(/*frame*/) { // you should override this method
  printf("update() method called...\n")
}

void Player::finish() { // you should override this method
  printf("finish() method called...\n")
}

void Player::run() {
  sendToServer("get_info");
  info = receive();
  init(info);
  sendToServer("ready");

  do {
    frame = receive();
    if (frame &&
        check_frame(/*json.loads(frame)*/)) // return false if we need to quit
      update(/*json.loads(frame)*/);
    else {
      finish();
      return;
    }
  }
}
